### ISSUE TITLE
Add support for preview features in PmdExtension

### ISSUE THREAD

**Issue by DavidOpDeBeeck** [gradle/gradle#21519]
Currently it is not possible to run the PMD task on code that uses java preview features and is compiled using `--enable-preview`. The PMD task will not fail, but will also not run the rules on those classes. The following error is displayed:  `Error during type resolution of field '{redacted}' in class {redacted} due to: java.lang.UnsupportedClassVersionError: Preview features are not enabled for {redacted} (class file version 61.65535). Try running with '--enable-preview'`.

PMD supports a way to enable preview features by defining the java version as documented [here](https://pmd.github.io/latest/pmd_languages_java.html).

### Expected Behavior
The PMD task runs the rules on java files that use preview features.

### Current Behavior
The PMD task fails to run the rules on java files that use preview features.

### Context
We currently can not rely on PMD to verify our rules on classes where we use preview features.

### Related issues

#11107 and #15280.

### Proposed change

1. Adding a new field `Propery<String> javaVersion` to [PmdExtension.java](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/PmdExtension.java) 
2. Adding a mapping from the `javaVersion` field to the `sourceLanguage` from `PMDTask` in [PmdInvoker.groovy](https://github.com/gradle/gradle/blob/master/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/PmdInvoker.groovy)
```
ant.pmd(antPmdArgs) {
  sourceLanguage(name: "java", version: javaVersion)
}
``` 

**Note**: There currently exists a `TargetJdk` enum that is used for PMD versions < 5.0.0. I would not use/extend that enum with more versions, because PMD only supports the preview language features for the latest two java versions and that would mean updating that enum for each java release or when PMD changes it's mind. Keeping it a `String` is the most flexible solution.

**Comment by DavidOpDeBeeck** [comment]
Or maybe make the source language fully customizable by defining a `languageName` and a `languageVersion`? The language name can default to `java`, but defining a default for the version is not possible I think.

**Comment by DavidOpDeBeeck** [comment]
@JD-CSTx 

PMD needs 2 things to correctly analyse java code that uses preview features:
* `--enable-preview` needs to set, either using `PMD_JAVA_OPTS` or passing it to the jvm running the PMD code
  * when this is not set PMD will log: `Error during type resolution of field '{redacted}' in class {redacted} due to: java.lang.UnsupportedClassVersionError: Preview features are not enabled for {redacted} (class file version 61.65535). Try running with '--enable-preview'`
* a language version with the `-preview` suffix needs to be selected.
  * when this is not set PMD will log: `Caused by: net.sourceforge.pmd.lang.java.ast.ParseException: Line 24, Column 73: Pattern Matching in Switch is only supported with JDK 17 Preview.` (an example of a preview feature)

Your solution fixes the first problem, but I don't see how it allows for PMD to parse preview features.

**Comment by JD-CSTx** [comment]
> Your solution fixes the first problem, but I don't see how it allows for PMD to parse preview features.
@DavidOpDeBeeck You're right, in the end it doesn't fix anything at all. We really need an "enable-preview" feature.

**Comment by ov7a** [comment]
Sorry for the late reply.

This issue needs a decision from the team responsible for that area. They have been informed. Response time may vary.

**Comment by big-guy** [comment]
We now use worker processes to run PMD, so it seems like a fix here would need to:
1. Take into account if the code being analyzed enabled feature previews
2. Add `--enable-preview` to the worker's arguments
3. Pass any additional information to PMD so it knows which language level to expect

It would be nice if the information about enabling feature previews was available from the compilation step in a cross cutting way. IOW, the PMD plugin shouldn't have to figure this out on its own, it should be known for the entire source set. 

PMD is not an area of focus right now, so we can't provide a timeline of when this would be fixed.
